README.md
[<img src="https://f-droid.org/badge/get-it-on.png" alt="Get it on F-Droid" height="80">](https://f-droid.org/app/de.marmaro.krt.ffupdater)

## Firefox-Updater

Download, install and update these browsers from Mozilla:
 - [Firefox Browser / Fennec Release](https://play.google.com/store/apps/details?id=org.mozilla.firefox)
 - [Firefox Focus](https://play.google.com/store/apps/details?id=org.mozilla.focus)
 - [Firefox Klar](https://play.google.com/store/apps/details?id=org.mozilla.klar)
 - [Firefox Lite](https://play.google.com/store/apps/details?id=org.mozilla.rocket)
 - [Firefox Preview / Fenix](https://play.google.com/store/apps/details?id=org.mozilla.fenix)

FFUpdater will check periodically for updates and will display a notification when an update is available. This feature itself can be disabled and the check frequency can be changed.

Security measures:
 - only HTTPS connections
 - check fingerprint of the downloaded file
 - check fingerprint of the installed app

You can find the APK certificate fingerprints on multiple website - e.g. apkmirror.com
I did my best to make the app as secure as possible - feel free to double-check it in the source code.

Important: I removed Fennec Beta and Fennec Nightly because they are outdated - I will add Fenix Beta and Fenix Nightly soon.

## Deprecated browsers

### Fennec
16.05.2020: Mozilla wants to migrate from Fennec to Fenix. Fennec Beta and Fennec Nightly are already end-of-life and Fennec Release will be soon.

>Fennec is being replaced by our new state-of-the-art mobile browser codenamed "Fenix". We're slowly migrating users in order to make sure the experience is as painless and as enjoyable as possible. We started to migrate users who were using Fennec Nightly in January (bug 1608882). It took us several weeks to be sure of the result and to finally offer Fenix Nightly to all users using Fennec Nightly. Another few weeks later, we repeated the same process with Fennec Beta (bug 1614287). Fenix Beta has been offered to the whole Fennec Beta population on April 22nd. We're planning to do the same with Fennec Release sometimes this year. The schedule is still to be determined.

 >The Google Play Store[1] has a lot of nice features, but it's still pretty basic whenever a software publisher wants to slowly migrate users. Once a migration is started, we can't provide any Fennec updates to the population who wasn't offered Fenix, yet. I can say this restriction is painful to manage for Android developers, Mozilla included. Because of it, we had to stop shipping Fennec Nightly/Beta APKs at the beginning of each migration. This explains the dates of the last builds. At the same time, we stopped building Fennec Nightly/Beta because it enabled us to save technical resources[2] as well as people's time[3].

https://bugzilla.mozilla.org/show_bug.cgi?id=1627518

## Build app

### Android Studio

#### From an existing project

 1. File
 2. New
 3. Project from Version Control...
 4. Git
 5. URL: https://notabug.org/Tobiwan/ffupdater.git
 6. Clone
 7. Wait for Android Studio to import project
 8. Build
 9. Build Bundle(s) / APK(s)
 10. Build APK(s)
 11. Wait
 12. Click in the bottom right on "locate"

#### From "Welcome to Android Studio" screen

 1. Check out project from Version Control
 2. Git
 3. URL: https://notabug.org/Tobiwan/ffupdater.git
 4. Clone
 5. "Would you like to open it?" Yes
 6. Wait for Android Studio to import project
 7. Build
 8. Build Bundle(s) / APK(s)
 9. Build APK(s)
 10. Wait
 11. Click in the bottom right on "locate"

### Gradle

 1. `git clone https://notabug.org/Tobiwan/ffupdater.git`
 2. `cd ffupdater`
 3. create the file `local.properties` with the path to your Android SDK - for example:
 ````
 sdk.dir=C\:\\Users\\Tobias\\AppData\\Local\\Android\\Sdk
 ````
 4. `gradlew assembleDebug`
 5. `cd ffupdater/build/outputs/apk/debug`
 6. `ffupdater-debug.apk` is the generated APK file


## Mirrors
 - [Notabug.org](https://notabug.org/Tobiwan/ffupdater) ([Account](https://notabug.org/Tobiwan))
 - [Github](https://github.com/Tobi823/ffupdater) ([Account](https://github.com/Tobi823))
 - [Gitlab](https://gitlab.com/Tobiwan/ffupdater) ([Account](https://gitlab.com/Tobiwan))

## Technical detail

### Fennec Release

The Mozilla API will be queried:
 - https://archive.mozilla.org/pub/mobile/releases/ for latest version names
 - https://download.mozilla.org/?product=%s&os=%s&lang=multi to download the  app
   - first parameter: fennec-latest, fennec-beta-latest or fennec-nightly-latest
   - second parameter: android or android-x86

### Firefox Focus, Firefox Klar, Firefox Lite, Fenix

The Github API will be queried:
 - https://api.github.com/repos/mozilla-mobile/focus-android/releases/latest for Firefox Focus and Firefox Klar
 - https://api.github.com/repos/mozilla-tw/FirefoxLite/releases/latest for Firefox Lite
 - https://api.github.com/repos/mozilla-mobile/fenix/releases/latest for Fenix
 
If .../releases/latest returns a release without assets (APK files for download), then fetch all releases and search
for the latest release with assets.

## History

Since Mozilla [shut down their FTP server on 2015-08-05](https://blog.mozilla.org/it/2015/07/27/product-delivery-migration-what-is-changing-when-its-changing-and-the-impacts/), non-"Google Play Store" updates of Firefox have to be done by third-party apps such as this one.


Mozilla now uses a uniform URL to point to the latest release, see [their README](https://archive.mozilla.org/pub/mobile/releases/latest/README.txt) for details. The current URL for Android is: https://download.mozilla.org/?product=fennec-latest&os=android&lang=multi

I opened a ticket about non-"Google Play Store" updates with Mozilla in 2015, as did others:

- [ ] https://bugzilla.mozilla.org/show_bug.cgi?id=1192279
- [x] https://bugzilla.mozilla.org/show_bug.cgi?id=1220773

### Maintainer:

#### Tobiwan (now)

#### Boris Kraut (https://gitlab.com/krt/ffupdater, until April 2019)
> Since I left F-Droid (and Android/Smartphones) about a year ago, I am looking for a new maintainer to take over. Unfortunately the upstream issue I opened years ago is still not solved in 2019. While Fennec F-Droid is back in the mainline repo and other binary repos do serve Firefox, some might still prefer this updater. So as I said: Maintainers welcome. The main task should be to test the last few merge requests (especially the background update stuff) and release a new version.
> **New Maintainer: https://notabug.org/Tobiwan/ffupdater**

## License

````
FFUpdater -- a simple Android app to update Firefox.
Copyright (C) 2015-2019 Boris Kraut <krt@nurfuerspam.de>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.
````


NEWS
??? Version 68.3 (35)
 * Color of collapsed title will be always white (thanks DeenHyper74)

2020-05-13 Version 68.2 (34)
 * Fix Fenix download from Github (thanks yhoyhoj)
 * Show correct download progress when downloading an app

2020-05-13 Version 68.1 (33)
 * Thanks DeenHyper74 for the Russian translation
 * Add support for Dark Theme (thanks DeenHyper74 for the tip)
 * Add setting for switching between Dark and Light Theme

2020-05-06 Version 68.0 (32)
 * Really big update
 * Add support for Firefox Klar, Firefox Focus, Firefox Lite and Fenix
 * Download and install the app inside FFUpdater (thanks wolfgang42 for the groundwork)
 * Improve UI
 * Verify the certificate of the downloaded and installed app
 * Many improvements
 * Thanks DeenHyper74 and xin for translations
 * Remove Fennec Beta and Fennec Nightly because their are not developed anymore https://bugzilla.mozilla.org/show_bug.cgi?id=1627518

2019-06-28 Version 67.4 (31)
 * Fix warning dialogue disappears after screen rotation (thanks DeenHyper74)

2019-06-28 Version 67.3 (30)
 * Fix crash when selecting an entry after rotating the channel dialog (thanks DeenHyper74)

2019-06-28 Version 67.2 (29)
 * Fix crash when rotating on channel dialog (thanks DeenHyper74)

2019-06-28 Version 67.1 (28)
 * Update Russian translation (thanks DeenHyper74)

2019-06-28 Version 67.0 (27)
 * Fix broken nightly download (thanks dannycolin for the info)
 * Display warning when switching from 'Release' channel to the 'Nightly' or 'Beta' channel (thanks DeenHyper74)
 * Interval between update checks is now configurable (thanks aplufr, wah6Me1l and DeenHyper74)

2019-04-25 Version 66.2 (26)
 * Improve french translation (thanks xinxinxinxinxin)
 * Fix "update notification will be shown every 5 minutes" (bug discovered by aplufr)

2019-04-25 Version 66.1 (25)
 * Add grammar fixes (thanks DeenHyper74)

2019-04-25 Version 66.0 (24)
 * Notification (for a Firefox update) now works on Android 9
 * Replace BackgroundService with WorkManager (AndroidX) for requesting Mozilla's API
 * Clean up code (thanks DeenHyper74)
 * Fix some minor bugs (thanks DeenHyper74)
 * App requires API level 18 because Firefox requires API level 18
 * Update translation

2019-04-01 Version 65.0 (23)
 * Added support for beta and nightly channels
 * Switched to light theme
 * Update russian translation
 * Handover maintainership to https://notabug.org/Tobiwan/ffupdater

2018-01-07 Version 57.0 (22)
 * Add some translations

2017-06-13 Version 54.0 (21)
 * Add license report
 * Update icon
 * Add some translations

2017-05-20 Version 53.0 (20)
 * Update to 53.0
 * Use new Mozilla API to det version information
 * Remove a lot of unused code.
 * Add icon
 * Fix crash with SDK < 17
 * Enable smaller builds

2017-01-22 Version 51.0 (18)
 * Update to 51.0
 * Remove a lot of unused code.
 * Re-implement actual checking.

2016-04-29 Version 46.0 (16)
 * No changes, just bump to remind people

2016-04-06 Version 45.0.1 (15)
 * Remove everything but URL generator and download button.

2016-03-09 Version 45.0a (14)
 * Really, really quickfix Android6 issues..
 * Remove version check since it's broken: Just download the APK.

2016-03-06 Version 45.0 (13)
 * Use lower target to quickfix Android6 permissions

2016-01-26 Version 44.0 (12)
 * Update to reflect new Firefox version, but no change in
   architecture. Mozilla "-latest" URL still works...

2015-12-31 Version 43.0.x (11)
 * Mozilla removed /latest/* downloads from archive. As recommended in
   https://archive.mozilla.org/pub/mobile/releases/latest/README.txt
   we use https://download.mozilla.org/?product=fennec-latest now.

2015-12-25 Version 43.0 (10)
 * Update to 43.0
 * Mozilla does not populate /latest/ anymore, see
   https://bugzilla.mozilla.org/show_bug.cgi?id=1233399

2015-12-24 Version 42.0.2 (9)
 * Update to 42.0.2

2015-11-23 Version 42.0.1 (8)
 * Update to 42.0.1

2015-11-02 Version 42.0 (7)
 * Update to 42.0

2015-09-23 Version 41.0 (6)
 * Bump to 41.0 (6)

2015-09-01 Version 40.0.3 (5)
 * Mark background setting as non-functional for now.
 * Update to 40.0.3

2015-08-10 Version 40.0 (3)
 * Remove FTP lookup since Mozilla is shutting down the servers.
 * Handle both request type by a single button.
 * "I am feeling lucky" now gets the next release, not the latest.
 * Add preferences to restrict connections: WiFi-only, metered, roaming.
 * Use actionbar.

2015-07-08 Version 39.0 (1)
 * Select download uri based on arch and api.
 * Check filename via FTP.
 * Download update file via HTTPS and DownloadManager.
 * Log errors and status.
 * Toast on updates.
 * Prompt for update (if necessary).
 * Option to use fixed/tested download location.
 * Add proper LICENSE (GPLv3+).


ffupdater/src/main/java/de/marmaro/krt/ffupdater/App.java
package de.marmaro.krt.ffupdater;

import android.content.Context;

import androidx.annotation.NonNull;

import org.apache.commons.codec.binary.ApacheCodecHex;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import de.marmaro.krt.ffupdater.device.DeviceEnvironment;
import de.marmaro.krt.ffupdater.device.DeviceEnvironment.ABI;

import static de.marmaro.krt.ffupdater.device.DeviceEnvironment.ABI.X86;
import static de.marmaro.krt.ffupdater.device.DeviceEnvironment.ABI.X86_64;

/**
 * All supported apps.
 * You can verify the APK certificate fingerprint on <a href="https://www.apkmirror.com">APKMirror</a> and other sites.
 */
public enum App {
    @SuppressWarnings("SpellCheckingInspection")
    FENNEC_RELEASE(R.string.fennec_release_title,
            R.string.fennec_release_description,
            R.string.empty,
            R.string.mozilla,
            "org.mozilla.firefox",
            "a78b62a5165b4494b2fead9e76a280d22d937fee6251aece599446b2ea319b04",
            Collections.emptyList(),
            16),
    @SuppressWarnings("SpellCheckingInspection")
    FIREFOX_KLAR(R.string.firefox_klar_title,
            R.string.firefox_klar_description,
            R.string.empty,
            R.string.github,
            "org.mozilla.klar",
            "6203a473be36d64ee37f87fa500edbc79eab930610ab9b9fa4ca7d5c1f1b4ffc",
            Arrays.asList(X86, X86_64),
            21),
    @SuppressWarnings("SpellCheckingInspection")
    FIREFOX_FOCUS(R.string.firefox_focus_title,
            R.string.firefox_focus_description,
            R.string.empty,
            R.string.github,
            "org.mozilla.focus",
            "6203a473be36d64ee37f87fa500edbc79eab930610ab9b9fa4ca7d5c1f1b4ffc",
            Arrays.asList(X86, X86_64),
            21),
    FIREFOX_LITE(R.string.firefox_lite_title,
            R.string.firefox_lite_description,
            R.string.empty,
            R.string.github,
            "org.mozilla.rocket",
            "863a46f0973932b7d0199b549112741c2d2731ac72ea11b7523aa90a11bf5691",
            Collections.emptyList(),
            21),
    @SuppressWarnings("SpellCheckingInspection")
    FENIX(R.string.fenix_title,
            R.string.fenix_description,
            R.string.fenix_warning,
            R.string.github,
            "org.mozilla.fenix",
            "5004779088e7f988d5bc5cc5f8798febf4f8cd084a1b2a46efd4c8ee4aeaf211",
            Collections.emptyList(),
            21);

    private final int titleId;
    private final int descriptionId;
    private final int warningId;
    private final int downloadSourceId;
    private final String packageName;
    private final byte[] signatureHash;
    private final List<ABI> unsupportedAbis;
    private final int minApiLevel;

    App(int titleId, int descriptionId, int warningId, int downloadSourceId, String packageName, String signatureHash, List<ABI> unsupportedAbis, int minApiLevel) {
        this.titleId = titleId;
        this.descriptionId = descriptionId;
        this.warningId = warningId;
        this.downloadSourceId = downloadSourceId;
        this.packageName = packageName;
        this.signatureHash = ApacheCodecHex.decodeHex(signatureHash);
        this.unsupportedAbis = unsupportedAbis;
        this.minApiLevel = minApiLevel;
    }

    @NonNull
    public String getTitle(Context context) {
        return context.getString(titleId);
    }

    @NonNull
    public String getDescription(Context context) {
        return context.getString(descriptionId);
    }

    @NonNull
    public String getWarning(Context context) {
        return context.getString(warningId);
    }

    @NonNull
    public String getDownloadSource(Context context) {
        return context.getString(downloadSourceId);
    }

    @NonNull
    public String getPackageName() {
        return packageName;
    }

    /**
     * @return SHA-256 hash of the APK signature
     */
    @NonNull
    public byte[] getSignatureHash() {
        return signatureHash;
    }

    public int getMinApiLevel() {
        return minApiLevel;
    }

    /**
     * @param deviceABI device metadata
     * @return can the app be installed on the device?
     */
    public boolean isCompatibleWithDevice(DeviceEnvironment deviceABI) {
        return !isIncompatibleWithDeviceAbi(deviceABI) && !isIncompatibleWithDeviceApiLevel(deviceABI);
    }

    /**
     * @param deviceABI device metadata
     * @return is the app not available for the device's ABI?
     */
    public boolean isIncompatibleWithDeviceAbi(DeviceEnvironment deviceABI) {
        return unsupportedAbis.contains(deviceABI.getBestSuitedAbi());
    }

    /**
     * @param deviceABI device metadata
     * @return is the app not compatible with the device's API Level?
     */
    public boolean isIncompatibleWithDeviceApiLevel(DeviceEnvironment deviceABI) {
        return !deviceABI.isSdkIntEqualOrHigher(minApiLevel);
    }
}
